data/transactions/logic/eval.go

```diff
@@ -4580,64 +4611,88 @@ func (cx *EvalContext) localsReference(account stackValue, ref uint64) (basics.A
 	return addr, app, addrIdx, nil
 }
 
-func (cx *EvalContext) assetReference(ref uint64, foreign bool) (basics.AssetIndex, error) {
+func (cx *EvalContext) assetReference(ref uint64, foreign bool) (aid basics.AssetIndex, err error) {
 	if cx.version >= directRefEnabledVersion {
-		aid := basics.AssetIndex(ref)
-		if cx.availableAsset(aid) {
-			return aid, nil
-		}
+		return cx.resolveAsset(ref)
+	}
 
-		if cx.version < sharedResourcesVersion {
-			// Allow use of indexes, but this comes last so that clear advice can be
-			// given to anyone who cares about semantics in the first few rounds of
-			// a new network - don't use indexes for references, use the asa ID.
-			if ref < uint64(len(cx.txn.Txn.ForeignAssets)) {
-				return basics.AssetIndex(cx.txn.Txn.ForeignAssets[ref]), nil
+	// resolveAsset is already similarly protected (and must be, since it is
+	// called independently)
+	if cx.Proto.AppForbidLowResources {
+		defer func() {
+			if aid <= lastForbiddenResource && err == nil {
+				err = fmt.Errorf("low Asset lookup %d", aid)
 			}
+		}()
+	}
+	// Old rules
+	if foreign {
+		// In old versions, a foreign reference must be an index in ForeignAssets
+		if ref < uint64(len(cx.txn.Txn.ForeignAssets)) {
+			return basics.AssetIndex(cx.txn.Txn.ForeignAssets[ref]), nil
 		}
-	} else {
-		// Old rules
-		if foreign {
-			// In old versions, a foreign reference must be an index in ForeignAssets
-			if ref < uint64(len(cx.txn.Txn.ForeignAssets)) {
-				return basics.AssetIndex(cx.txn.Txn.ForeignAssets[ref]), nil
+		return 0, fmt.Errorf("Asset index %d beyond txn.ForeignAssets", ref)
+	}
+	// Otherwise it's direct
+	return basics.AssetIndex(ref), nil
+}
+
+const lastForbiddenResource = 255
+
+// resolveAsset figures out what Asset an integer is referring to, considering 0 as
+// current app first, then uses the integer as is if it is an availableAsset, then
+// tries to perform a slot lookup.
+func (cx *EvalContext) resolveAsset(ref uint64) (aid basics.AssetIndex, err error) {
+	if cx.Proto.AppForbidLowResources {
+		defer func() {
+			if aid <= lastForbiddenResource && err == nil {
+				err = fmt.Errorf("low Asset lookup %d", aid)
 			}
-		} else {
-			// Otherwise it's direct
-			return basics.AssetIndex(ref), nil
-		}
+		}()
+	}
+	aid = basics.AssetIndex(ref)
+	if cx.availableAsset(aid) {
+		return aid, nil
 	}
-	return 0, fmt.Errorf("invalid Asset reference %d", ref)
 
+	// Allow use of indexes, but this comes last so that clear advice can be
+	// given to anyone who cares about semantics in the first few rounds of
+	// a new network - don't use indexes for references, use the Asset ID
```

Comment:
could you eleborate more on the clear advice you are referring here... plz T_T?

Reply:
The advice is, "don't use indexes for references, use the Asset ID".

Which is now in the documentation as:
```
It is recommended that contract authors avoid supplying array indexes
to these opcodes, and always use explicit resource IDs. By using
explicit IDs, contracts will better take advantage of group resource
sharing.  The array indexing interpretation may be deprecated in a
future version.
```